File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.cpp
class Solution {
public:
    std::vector<int> helper(TreeNode *root, std::vector<int>& path) {
        if (root == nullptr) return path;
        helper(root->left, path);
        path.push_back(root->val);
        helper(root->right, path);
        return path;
    }
    std::vector<int> inorderTraversal(TreeNode *root) {
        std::vector<int> path;
        helper(root, path);
        return path;
    }
};

File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        fn inorder(node: &Option<Rc<RefCell<TreeNode>>>, result: &mut Vec<i32>) {
            if let Some(v) = node {
                let v = v.borrow();
                inorder(&v.left, result);
                result.push(v.val);
                inorder(&v.right, result);
            }
        }
        let mut result = vec![];
        if let Some(v) = root {
            inorder(&Some(v), &mut result);
        }
        result
    }
}
